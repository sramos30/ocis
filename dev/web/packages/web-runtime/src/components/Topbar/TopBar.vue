<template>
  <header
    id="oc-topbar"
    class="oc-flex oc-flex-middle oc-flex-between oc-px-m"
    :aria-label="$gettext('Top bar')"
  >
    <div class="oc-topbar-left oc-flex oc-flex-middle oc-flex-between">
      <applications-menu v-if="appMenuItems.length" :applications-list="appMenuItems" />
      <router-link ref="navigationSidebarLogo" to="/">
        <oc-img :src="logoImage" :alt="sidebarLogoAlt" class="oc-logo-image" />
      </router-link>
    </div>
    <div class="portal-wrapper">
      <portal-target name="app.runtime.header" multiple></portal-target>
    </div>
    <div class="oc-topbar-right oc-flex oc-flex-middle oc-flex-between">
      <theme-switcher v-if="darkThemeAvailable" />
      <feedback-link v-if="isFeedbackLinkEnabled" />
      <notifications v-if="isNotificationBellEnabled" />
      <user-menu v-if="isUserMenuEnabled" :applications-list="userMenuItems" />
    </div>
  </header>
</template>

<script>
import { mapGetters } from 'vuex'
import NavigationMixin from '../../mixins/navigationMixin'

import ApplicationsMenu from './ApplicationsMenu.vue'
import UserMenu from './UserMenu.vue'
import Notifications from './Notifications.vue'
import FeedbackLink from './FeedbackLink.vue'
import ThemeSwitcher from './ThemeSwitcher.vue'

export default {
  components: {
    ApplicationsMenu,
    FeedbackLink,
    Notifications,
    ThemeSwitcher,
    UserMenu
  },
  mixins: [NavigationMixin],
  props: {
    applicationsList: {
      type: Array,
      required: false,
      default: () => []
    },
    activeNotifications: {
      type: [Array, Boolean],
      required: false,
      default: () => []
    }
  },
  computed: {
    ...mapGetters(['configuration', 'user']),

    activeRoutePath() {
      return this.$router.resolve(this.$route).location.path
    },

    appMenuItems() {
      return this.navigation_getMenuItems([null, 'apps', 'appSwitcher'], this.activeRoutePath)
    },
    userMenuItems() {
      return this.navigation_getMenuItems(['user'], this.activeRoutePath)
    },

    darkThemeAvailable() {
      return this.configuration.themes.default && this.configuration.themes['default-dark']
    },

    sidebarLogoAlt() {
      return this.$gettext('Navigate to personal files page')
    },

    logoImage() {
      return this.configuration.currentTheme.logo.topbar
    },

    isFeedbackLinkEnabled() {
      return !this.configuration.options.disableFeedbackLink
    },

    isNotificationBellEnabled() {
      return this.user?.id && this.activeNotifications.length
    },

    isUserMenuEnabled() {
      return this.user?.id
    }
  }
}
</script>

<style lang="scss">
#oc-topbar {
  height: 60px;
  position: sticky;
  z-index: 5;

  img {
    max-height: 42px;
    image-rendering: auto;
    image-rendering: crisp-edges;
    image-rendering: pixelated;
    image-rendering: -webkit-optimize-contrast;
    user-select: none;
  }

  .oc-topbar-left {
    gap: 30px;

    img.oc-logo-image {
      height: 42px;
    }
  }
  .oc-topbar-right {
    gap: 20px;
  }
  @media only screen and (max-width: 960px) {
    .oc-topbar-left,
    .oc-topbar-right {
      gap: 0.5rem;
    }
  }
}
</style>
