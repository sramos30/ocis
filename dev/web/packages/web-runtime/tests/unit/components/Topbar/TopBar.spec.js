import { shallowMount, createLocalVue } from '@vue/test-utils'
import Vuex from 'vuex'
import VueRouter from 'vue-router'
import TopBar from 'web-runtime/src/components/Topbar/TopBar.vue'
import stubs from '../../../../../../tests/unit/stubs'

const localVue = createLocalVue()
localVue.use(Vuex)
localVue.use(VueRouter)

jest.spyOn(TopBar.mixins[0].methods, 'navigation_getMenuItems').mockImplementation(() => [
  {
    icon: 'folder',
    iconUrl: undefined,
    title: 'Files',
    path: '/files'
  },
  {
    icon: 'some-icon',
    iconUrl: undefined,
    title: 'External',
    url: 'http://some.org',
    target: '_blank'
  },
  {
    icon: 'application',
    iconUrl: undefined,
    path: '/settings',
    title: 'Settings'
  }
])

describe('Top Bar component', () => {
  it('Displays applications menu', () => {
    const wrapper = shallowMount(TopBar, {
      store: new Vuex.Store({
        getters: {
          configuration: () => feedbackButtonPresent(true),
          user: () => ({
            id: 'einstein'
          })
        }
      }),
      localVue,
      router: new VueRouter(),
      stubs,
      propsData: {
        applicationsList: ['testApp']
      }
    })

    expect(wrapper.html().indexOf('applications-menu-stub')).toBeGreaterThan(-1)
    expect(wrapper).toMatchSnapshot()
  })
})

const feedbackButtonPresent = (enabled) => ({
  options: { disableFeedbackLink: !enabled },
  themes: {
    default: {},
    'default-dark': {}
  },
  currentTheme: {
    logo: {
      topbar: 'example-logo.svg'
    }
  }
})
