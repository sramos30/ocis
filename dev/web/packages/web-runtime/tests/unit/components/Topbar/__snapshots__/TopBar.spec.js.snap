// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Top Bar component Displays applications menu 1`] = `
<header id="oc-topbar" aria-label="Top bar" class="oc-flex oc-flex-middle oc-flex-between oc-px-m">
  <div class="oc-topbar-left oc-flex oc-flex-middle oc-flex-between">
    <applications-menu-stub applicationslist="[object Object],[object Object],[object Object]"></applications-menu-stub>
    <router-link-stub to="/" tag="a" ariacurrentvalue="page" event="click">
      <oc-img-stub src="example-logo.svg" alt="Navigate to personal files page" class="oc-logo-image"></oc-img-stub>
    </router-link-stub>
  </div>
  <div class="portal-wrapper">
    <portal-target-stub name="app.runtime.header" multiple=""></portal-target-stub>
  </div>
  <div class="oc-topbar-right oc-flex oc-flex-middle oc-flex-between">
    <theme-switcher-stub></theme-switcher-stub>
    <feedback-link-stub></feedback-link-stub>
    <!---->
    <user-menu-stub applicationslist="[object Object],[object Object],[object Object]"></user-menu-stub>
  </div>
</header>
`;
